pkg/utils/managedresources/builder/secrets.go

```diff
@@ -47,9 +52,45 @@ func (s *Secret) WithNamespacedName(namespace, name string) *Secret {
 	return s
 }
 
-// WithLabels sets the labels.
+// WithLabels sets the labels. The label "resources.gardener.cloud/garbage-collectable-reference" is retained
+// if it already exists in the current labels.
 func (s *Secret) WithLabels(labels map[string]string) *Secret {
-	s.secret.Labels = labels
+	if s.secret.Labels == nil {
+		s.secret.Labels = cloneLabels(labels)
+		return s
+	}
+
+	_, ok := s.secret.Labels[references.LabelKeyGarbageCollectable]
+	if ok && pointer.BoolDeref(s.secret.Immutable, false) {
+		s.secret.Labels = make(map[string]string, len(labels)+1)
+		s.secret.Labels[references.LabelKeyGarbageCollectable] = references.LabelValueGarbageCollectable
+		for k, v := range labels {
+			s.secret.Labels[k] = v
+		}
+		return s
+	}
+
+	s.secret.Labels = cloneLabels(labels)
+	return s
+}
```

Comment:
```suggestion
// WithLabels sets the labels. The label "resources.gardener.cloud/garbage-collectable-reference" is retained
// if it already exists in the current labels.
func (s *Secret) WithLabels(labels map[string]string) *Secret {
	if s.secret.Labels == nil {
		s.secret.Labels = utils.MergeStringMaps(labels)
		return s
	}

	_, ok := s.secret.Labels[references.LabelKeyGarbageCollectable]
	if ok && pointer.BoolDeref(s.secret.Immutable, false) {
		s.secret.Labels = map[string]string{references.LabelKeyGarbageCollectable: references.LabelValueGarbageCollectable}
	}

	s.secret.Labels = utils.MergeStringMaps(labels, s.secret.Labels)
	return s
}
```

We already use `utils.MergeStringMaps` to copy maps in other places.